### PR TITLE
TE-5102: Git hook incorrect ruleset path

### PR DESCRIPTION
**Pull request by artem-png**
- Developer(s): @artem-png 

- Ticket: https://spryker.atlassian.net/browse/TE-5102

- Academy PR: ACADEMY_URL_HERE

- Release Group: https://release.spryker.com/release-groups/view/2100

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   GitHook               | patch (currently 0.0.x)  |                       |

#### Release Notes

{skip}

-----------------------------------------

##### Change log

Bugfix

- Hardcoded ruleset.xml file for phpmd can be overridden by passing `rulesetPath` in config.

### CODE REVIEW

File: src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php [hunk 1]
```diff
@@ -34,4 +36,34 @@ public function getMinimumPriority(): int
...
+            'vendor/spryker/development/src/Spryker/Zed/Development/Business/PhpMd/ruleset.xml',
+        ];
+
+        foreach ($defaultRuleSetPathsByPriority as $ruleSetPath) {
+            if (file_exists(PROJECT_ROOT . DIRECTORY_SEPARATOR . $ruleSetPath)) {
+                return $ruleSetPath;
+            }
+        }
+
+        throw new Exception('Ruleset file was not found in any location: ' . PHP_EOL . implode(PHP_EOL, $defaultRuleSetPathsByPriority));
```
Comment (limeeugenia) [inline src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php]: This look misplaced in the config, I'd rather have this logic in a model because it looks a lot like business logic to me.

Config should only have default paths as an array, then the detection and exception can move to the model.

Comment (limeeugenia) [inline src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php]: We usually try to avoid throwing the general `Exception` in favour of throwing the specific one.

File: src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php [hunk 2]
```diff
@@ -34,4 +36,34 @@ public function getMinimumPriority(): int
...
+     * @throws \Exception
+     *
+     * @return string
+     */
+    protected function getDefaultRulesetPath(): string
+    {
+        $defaultRuleSetPathsByPriority = [
+            'config/phpmd-ruleset.xml',
+            'vendor/spryker/spryker/Bundles/Development/src/Spryker/Zed/Development/Business/PhpMd/ruleset.xml',
+            'vendor/spryker/development/src/Spryker/Zed/Development/Business/PhpMd/ruleset.xml',
```
Comment (limeeugenia) [inline src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php]: Core usually has the config for only the split scenario. You can have a look at how the `DevelopmentConfig::INTERNAL_NAMESPACES_TO_PATH_MAPPING` is defined and overridden on the project level.

The comment `@project Only needed in Project, not in demoshop` is required in this case.

File: src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php [hunk 3]
```diff
@@ -70,7 +70,7 @@ public function run(CommandContextInterface $context): CommandResultInterface
         if (!$process->isSuccessful()) {
             $commandResult
                 ->setError(trim($process->getErrorOutput()))
-                ->setMessage(trim($process->getOutput()));
+                ->setMessage(trim($process->getOutput() ?: $process->getErrorOutput()));
```
Comment (limeeugenia) [inline src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php:73]: Why is this needed? `ErrorOutput` is already in the `Error`.

**Review by limeeugenia — CHANGES_REQUESTED** [review]

File: src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php (hunk 3 above)
Comment (artem-png) [reply src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php:73]: $process->getOutput() can return empty string. I get empty string when path to file is incorrect. So by this fix we don't have empty exception in console and we always know about details

File: src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php (hunk 2 above)
Comment (limeeugenia) [inline src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php]: Maybe this can be set on a class level so that it's possible to extend this. What do you think?

File: src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php (hunk 3 above)
Comment (limeeugenia) [reply src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php:73]: I still get the same error message when I revert this line. Are you sure this is useful?

Comment (artem-png) [reply src/GitHook/Command/FileCommand/PreCommit/PhpMdCheckCommand.php:73]: Screenshot 2020-03-30 at 13 10 23

Now it's OK because you see exception from PhpMdCheckConfiguration class. But when we have exception in vendor/bin/phpmd we have empty red line. With this fix we have readable message that ruleset file not found

**Review by limeeugenia — APPROVED** [review]

File: src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php [hunk 4]
```diff
@@ -7,13 +7,24 @@
...
      */
     protected $config;
 
+    /**
+     * @var string[]
+     */
+    protected $defaultRuleSetPathsByPriority = [
+        'config/phpmd-ruleset.xml',
+        'vendor/spryker/spryker/Bundles/Development/src/Spryker/Zed/Development/Business/PhpMd/ruleset.xml',
+        'vendor/spryker/development/src/Spryker/Zed/Development/Business/PhpMd/ruleset.xml',
```
Comment (dereuromark) [inline src/GitHook/Command/FileCommand/PreCommit/PhpMd/PhpMdCheckConfiguration.php:25]: The order is wrong from a project perspective.
